### PR TITLE
Feature/userService

### LINKED ISSUE #10: [User] implement service

### PR DESCRIPTION
**Pull request by ttkmw**
## Related Issue

resolve: #10 

## Description
userService

## Checklist

Thank you for your contribution.
Before submitting this PR, please make sure:

- [x] Test case
- [x] End of Work

### CODE REVIEW

File: server/src/port/adapter/service/blockchain/ethereum/web3/test/socket/config/web3.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
 export default {
     url: 'localhost:8545',
-    type: 'socket',
+    type: 'http',
```
Comment (hihiboss) [inline server/src/port/adapter/service/blockchain/ethereum/web3/test/socket/config/web3.ts]: Should the directory be changed `server/src/port/adapter/service/blockchain/ethereum/web3/test/socket/config/web3.ts` to `server/src/port/adapter/service/blockchain/ethereum/web3/test/http/config/web3.ts`??

File: server/src/port/persistence/repository/ghost.repository.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { DeepPartial, EntityRepository, Repository, SaveOptions } from 'typeorm';
+import { Ghost } from '../../../domain/ghost/ghost.entity';
+
+@EntityRepository(Ghost)
+export class GhostRepository extends Repository<Ghost> implements IGhostRepository {
+
+}
+
+interface IGhostRepository {
```
Comment (hihiboss) [inline server/src/port/persistence/repository/ghost.repository.ts]: Please divide the interface `IGhostRepository` and the implementation `GhostRepository` in `ghost.repository.ts`.

File: server/src/port/persistence/repository/user.repository.ts [hunk 3]
```diff
@@ -1,12 +1,19 @@
...
+
+    findByAddress(address: string) {
+        return this.findOne({address});
+    }
 }
 
-interface IGhostRepository {
-  findOne(id?: string): Promise<Ghost | undefined>;
-  save(entity: Ghost): Promise<Ghost>;
+interface IUserRepository {
```
Comment (hihiboss) [inline server/src/port/persistence/repository/user.repository.ts]: Please divide the interface `IUserRepository` and the implementation `UserRepository` in `user.repository.ts`.

**Review by hihiboss — APPROVED** [review]
LGTM :)

File: server/src/app/user/user.service.impl.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            throw new Error('address should be defined');
+        }
+
+        const userRetrieved = await this.userRepository.findByAddress(userDto.address);
+        if (userRetrieved !== undefined) {
+            throw new Error('address is already registered');
+        }
+
+        const user = new User(userDto.address);
+
```
Comment (junbeomlee) [inline server/src/app/user/user.service.impl.ts]: plz remove empty line!

File: server/src/app/user/user.service.impl.spec.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            }).compile();
+            service = module.get<UserServiceImpl>(UserServiceImpl);
+
+            expect(service).toBeDefined();
+        });
+    });
+    describe('#get()', () => {
+        const mockUserId = 1;
+        it('should return mockUser', async () => {
+            const mockRepository: UserRepository = mock(UserRepository);
```
Comment (junbeomlee) [inline server/src/app/user/user.service.impl.spec.ts]: plz rename mockUser -> user

File: server/src/app/user/user.service.impl.spec.ts [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+import {UserDto} from '../../domain/user/dto/user.dto';
+import {UserRepository} from '../../port/persistence/repository/user.repository.impl';
+
+const mockAddress = 'testAddress';
+const mockUser = new User(mockAddress);
+
+describe('UserServiceImpl', async () => {
+
+    let service: UserServiceImpl;
+    describe('dependency resolve', () => {
```
Comment (junbeomlee) [inline server/src/app/user/user.service.impl.spec.ts:19]: put mockAddress, mockUser inside `describe('UserServiceImpl', async () => {`

File: server/src/app/user/user.service.impl.spec.ts [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        it('should return mockUser', async () => {
+            const mockRepository: UserRepository = mock(UserRepository);
+            when(mockRepository.findByAddress(mockAddress)).thenReturn(undefined);
+            when(mockRepository.save(anything())).thenReturn(new Promise((resolve) => {
+                resolve(mockUser);
+            }));
+            const mockedRepositoryInst: UserRepository = instance(mockRepository);
+            service = new UserServiceImpl(mockedRepositoryInst);
+            const userDto = new UserDto(mockAddress);
+
```
Comment (junbeomlee) [inline server/src/app/user/user.service.impl.spec.ts:64]: i think
```
const mockRepository: UserRepository = mock(UserRepository);
when(mockRepository.findByAddress(mockAddress)).thenReturn(undefined); when(mockRepository.save(anything())).thenReturn(new Promise((resolve) => {
        resolve(mockUser);
}));
const mockedRepositoryInst: UserRepository = instance(mockRepository);
 service = new UserServiceImpl(mockedRepositoryInst);
```
duplicated above code can be moved to
```
beforeEach('set user service', async () =>{
~~~
})
```

Comment (ttkmw) [reply server/src/app/user/user.service.impl.spec.ts:64]: @junbeomlee 
i don't think so. 
```
when(mockRepository.findByAddress(mockAddress)).thenReturn(undefined); when(mockRepository.save(anything())).thenReturn(new Promise((resolve) => {
        resolve(mockUser);
}));
```
is only used in '#get()' / 'should return user'.
i think it's better to define only in '#get()' / 'should return user'.

**Comment by ttkmw** [comment]
@junbeomlee @hihiboss 
review pls

**Comment by junbeomlee** [comment]
LGTM!

**Pull request merged**